File: 494/targetSum.js
/*
https://leetcode.com/problems/target-sum/description/?envType=daily-question&envId=2024-12-26
*/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */

var depthFirstSearch_forArrays=function(nums, target, lengthOfNums, beginIndex, currentSum){
    if(beginIndex===lengthOfNums){
        if(currentSum===target) return 1;
        return 0;
    }
    let pickPlus=0, pickMinus=0;
    pickPlus+=depthFirstSearch_forArrays(nums, target, lengthOfNums, beginIndex+1, currentSum+nums[beginIndex]);
    pickMinus+=depthFirstSearch_forArrays(nums, target, lengthOfNums, beginIndex+1, currentSum+(nums[beginIndex]*-1));
    return pickPlus+pickMinus;
}

var findTargetSumWays = function(nums, target) {
    let lengthOfNums=nums.length, beginIndex=0, currentSum=0;
    return depthFirstSearch_forArrays(nums, target, lengthOfNums, beginIndex, currentSum);
};

File: 4/494/targetSum.cpp
/*
https://leetcode.com/problems/target-sum/description/?envType=daily-question&envId=2024-12-26
*/

class Solution {
public:

    int dfs(vector<int>&nums, int target, int sz, int curidx, int cursum){
        if(curidx==sz){
            if(cursum==target) return 1;
            return 0;
        }
        int pkplus=0, pkminus=0;
        pkplus+=dfs(nums, target, sz, curidx+1, cursum+nums[curidx]);
        pkminus+=dfs(nums, target, sz, curidx+1, cursum+(nums[curidx]*-1));
        return pkplus+pkminus;
    }

    int findTargetSumWays(vector<int>& nums, int target) {
        int sz=nums.size(), idx=0, cursum=0;
        return dfs(nums, target, sz, idx, cursum);
    }
};